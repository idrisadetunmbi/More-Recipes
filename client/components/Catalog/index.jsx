import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroll-component';
import { Link } from 'react-router-dom';

import Recipe from '../Recipe';
import { LoaderWithComponent } from '../reusables';
import './index.scss';
import { fetchRecipes, searchRecipes } from '../../actions/recipes';

/**
 * @export
 * @class Catalog
 * @extends {Component}
 */
export class Catalog extends Component {
  state = {
    searchTerm: '',
    searchMode: false,
  }

  /**
   * @returns {void}
   * @memberOf Catalog
   */
  componentWillUnmount() {
    // scroll to top of the page incase infinite scroll 'scrollTop button'
    // gets clicked
    window.scrollTo(0, 0);
  }

  /**
   * onClick handler for add recipe button
   * @returns {void}
   * @memberOf Catalog
   */
  addRecipeOnClick = () => {
    const { history } = this.props;
    if (!this.props.user.data.token) {
      history.push('/signin', {
        modal: true,
        previousLocation: history.location.pathname,
      });
      return;
    }
    history.push('/recipes/create', {
      modal: true,
      previousLocation: history.location.pathname,
    });
  }

  /**
   * handles enter key press when search input element is focused
   * @param {object} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  enterKeyPress = (event) => {
    if (event.keyCode === 13) {
      event.stopPropagation();
      const searchInput = document.querySelector('input');
      const searchTerm = searchInput.value;
      if (!searchTerm) {
        return;
      }
      this.setState({ searchMode: true, searchTerm });
      this.props.searchRecipes(searchTerm);
    }
  }

  /**
   * @param {event} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  handleClearSearchClick = (event) => {
    const searchInput = document.querySelector('input');
    event.stopPropagation();
    searchInput.value = '';
    searchInput.focus();
    event.target.style.display = 'none';
    this.setState({ searchMode: false });
  }

  /**
   * @param {event} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  handleSearchInputChange = (event) => {
    const { value } = event.target;
    const clearSearchButton = document.getElementById('clear-search');
    if (!value) {
      this.setState({ searchMode: false });
      clearSearchButton.style.display = 'none';
      return;
    }
    clearSearchButton.style.display = 'inline-block';
  }

  renderLoader = () => (
    <div className="infinite-scroll">
      <div className="preloader-wrapper small active">
        <div className="spinner-layer spinner-green-only">
          <div className="circle-clipper left">
            <div className="circle" />
          </div>
          <div className="gap-patch">
            <div className="circle" />
          </div>
          <div className="circle-clipper right">
            <div className="circle" />
          </div>
        </div>
      </div>
    </div>
  )

  /**
   * @param {Array} recipes - list of recipes to render
   *
   * @returns {JSX.Element} - list of recipe cards
   * @memberOf Catalog
   */
  renderRecipes = recipes =>
    recipes.map(recipe => (
      <Link key={recipe.id} to={`/recipes/${recipe.id}`}>
        <Recipe gridStyle="l4 s12 m6" recipe={recipe} />
      </Link>
    ))

  renderSearch = () =>
    (
      <div id="search-section" className="row">
        <div className="col offset-l3 l9 m12 s12">
          <input
            placeholder="Search Recipe"
            type="search"
            onFocus={() => { document.addEventListener('keydown', this.enterKeyPress); }}
            onBlur={() => { document.removeEventListener('keydown', this.enterKeyPress); }}
            onChange={this.handleSearchInputChange}
          />
          <i
            onClick={this.handleClearSearchClick}
            id="clear-search"
            className="material-icons"
          >
            cancel
          </i>
        </div>
      </div>
    )

  renderSuggestionSection = () => (
    <div className="col l3" id="suggestion-section">
      <h5>Today{'\''}s Combo</h5>
      <div className="divider" />
      <div className="recipe">
        <div className="card">
          <div className="card-image">
            <img alt="" src="images/sample_recipes/recipe0116-xl-farro-with-vinegar-glazed-sweet-potato-and-apples.jpg" />
          </div>
          <div className="card-content">
            <span className="card-title">Farrow with vinegar, glazed sweet potato and apples</span>
            <p>I am a very simple card. I am good at containing small&nbsp;
          bits of information. I am convenient because I require&nbsp;
          little markup to use effectively.
            </p>
          </div>
          <div className="card-action">
            <a href="#">This is a link</a>
          </div>
        </div>
      </div>
    </div>
  )

  renderAddRecipeButton = () => (
    <div className="fixed-action-btn">
      <a className="btn-floating btn-large pulse" onClick={this.addRecipeOnClick}>
        <i className="material-icons">add</i>
      </a>
    </div>
  )

  renderCatalogWithInfiniteScroll = () => {
    const scrollEndOnClick = (event) => {
      event.preventDefault();
      $('html, body').animate({
        scrollTop: $('#catalog-component').offset().top,
      }, 500);
    };
    const endMessage = () => (
      <h5 className="infinite-scroll">
        There are no more recipes.
        <a onClick={scrollEndOnClick}> Go to Top</a>
      </h5>
    );

    const { recipes } = this.props;
    return (
      <LoaderWithComponent
        showLoader={recipes.requestInitiated && !recipes.recipes.length}
        component={
          <InfiniteScroll
            next={this.props.fetchRecipes}
            hasMore={!recipes.fetchedAll}
            loader={this.renderLoader()}
            scrollThreshold={0.8}
            endMessage={endMessage()}
            style={{ position: 'relative' }}
          >
            {this.renderRecipes(recipes.recipes)}
          </InfiniteScroll>}
      />
    );
  }

  renderSearchResults = () => {
    const { searchTerm } = this.state;
    const searchResults = this.props.recipes
      .searchResults[searchTerm.toLowerCase()];
    const endMessage = <h5 className="infinite-scroll">There are no more recipes matching the search term</h5>;

    return (
      <LoaderWithComponent
        showLoader={!searchResults}
        component={
          searchResults && searchResults.results.length ?
            <InfiniteScroll
              next={() => this.props.searchRecipes(this.state.searchTerm)}
              hasMore={!searchResults.matchedAll}
              loader={this.renderLoader()}
              scrollThreshold={0.8}
              endMessage={endMessage}
              style={{ position: 'relative' }}
            >
              {
                this.renderRecipes(this.props.recipes.recipes.filter(recipe =>
                  searchResults.results.includes(recipe.id)))
              }
            </InfiniteScroll> :
            <h5>No results found for {this.state.searchTerm}</h5>
        }
      />
    );
  }

  /**
   * @returns {JSX.Element} JSX element representing the whole catalog page
   * @memberOf Catalog
   */
  render() {
    return (
      <div id="catalog-component">
        {this.renderSearch()}
        <div id="main-divider" className="divider" />
        <div className="row">
          {this.renderSuggestionSection()}
          <div className="col l9 s12">
            <h5>{this.state.searchMode ?
              `Search Results for ${this.state.searchTerm}` : 'Featured Recipes'}
            </h5>
            <div className="divider" id="gallery-before" />
            {
              this.state.searchMode ?
                this.renderSearchResults() :
                this.renderCatalogWithInfiniteScroll()
            }
          </div>
        </div>
        {this.renderAddRecipeButton()}
      </div>
    );
  }
}

Catalog.propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.array,
    fetchedAll: PropTypes.bool,
    requestInitiated: PropTypes.bool.isRequired,
    searchResults: PropTypes.shape().isRequired,
  }).isRequired,

  history: PropTypes.shape({
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }),
  }).isRequired,

  user: PropTypes.shape({
    data: PropTypes.shape({
      token: PropTypes.string,
    }),
  }).isRequired,

  fetchRecipes: PropTypes.func.isRequired,
  searchRecipes: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  user: state.user,
  recipes: state.recipes,
});

const mapDispatchToProps = dispatch => ({
  fetchRecipes: () => dispatch(fetchRecipes()),
  searchRecipes: searchTerm => dispatch(searchRecipes(searchTerm)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
