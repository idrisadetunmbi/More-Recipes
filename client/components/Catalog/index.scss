#catalog-component {
  #main-divider {
    margin-left: 3em;
    margin-right: 3em; 
    margin-top: 0; 
    display: block;
  }
  
  #search-section {
    margin: 3em;
    margin-bottom: 0;
    input {
      margin-left: -1em;
      border: 1px solid #9e9e9e;
      padding-left: .8em;
      font-size: 1em;
      font-family: 'Lato';
      border-radius: .25rem;
    };
    div {
      position: relative;
    }
    i {
      position: absolute;
      top: 0.7rem;
      right: 1.5rem;
      cursor: pointer;
      display: none;
    }
  };

  #suggestion-section {
    padding: 3em;
    margin-top: 1em;
    .divider {
      margin-bottom: 2em;
    }
  };

  h5 {
    font-family: 'Raleway';
    font-style: italic;
    font-size: 1.5rem;
    padding-left: .5rem;
    text-align: center;
  };

  #gallery-before {
    margin: 0 .7rem 1rem .5rem;
  }

  .infinite-scroll {
    position: absolute;
    bottom: -3rem;
    margin: 0 auto;
    text-align: center;
    left: 0;
    right: 0;    
  }
}