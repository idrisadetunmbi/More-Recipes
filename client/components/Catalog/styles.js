export const searchInput = {
  marginLeft: '-1em',
  border: '1px solid #9e9e9e',
  height: '2.5em',
  width: '65%',
  marginRight: '0',
  paddingLeft: '.8em',
  fontSize: '1.5em',
  fontFamily: 'Open Sans',
  fontStyle: 'italic',
};

export const searchInputEnclosingDiv = {
  margin: '3em',
  marginBottom: '0',
};
