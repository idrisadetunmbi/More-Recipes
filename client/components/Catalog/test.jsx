import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Catalog } from '../Catalog';

Enzyme.configure({ adapter: new Adapter() });

describe('<Catalog />', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<Catalog recipes={{ recipes: [] }} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('simulates click events', () => {
    const wrapper = mount(<Catalog recipes={{ recipes: [] }} user={{ data: { } }} history={{ location: { pathname: '' }, push: jest.fn() }} />);
    const addRecipeOnClick = jest.spyOn(wrapper.instance(), 'addRecipeOnClick');
    wrapper.instance().forceUpdate();
    wrapper.find('.btn-floating').simulate('click');
    expect(addRecipeOnClick).toHaveBeenCalled();
  });
});
