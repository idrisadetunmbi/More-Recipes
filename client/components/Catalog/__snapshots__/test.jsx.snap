// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog component matches snapshot 1`] = `
<div
  id="catalog-component"
>
  <div
    className="row"
    id="search-section"
  >
    <div
      className="col offset-l3 l9 m12 s12"
    >
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search Recipe"
        type="search"
      />
      <i
        className="material-icons"
        id="clear-search"
        onClick={[Function]}
      >
        cancel
      </i>
    </div>
  </div>
  <div
    className="divider"
    id="main-divider"
  />
  <div
    className="row"
  >
    <div
      className="col l3"
      id="suggestion-section"
    >
      <h5>
        Today
        '
        s Combo
      </h5>
      <div
        className="divider"
      />
      <div
        className="recipe"
      >
        <div
          className="card"
        >
          <div
            className="card-image"
          >
            <img
              alt=""
              src="images/sample_recipes/recipe0116-xl-farro-with-vinegar-glazed-sweet-potato-and-apples.jpg"
            />
          </div>
          <div
            className="card-content"
          >
            <span
              className="card-title"
            >
              Farrow with vinegar, glazed sweet potato and apples
            </span>
            <p>
              I am a very simple card. I am good at containing small  bits of information. I am convenient because I require  little markup to use effectively.
            </p>
          </div>
          <div
            className="card-action"
          >
            <a
              href="#"
            >
              This is a link
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col l9 s12"
    >
      <h5>
        Featured Recipes
      </h5>
      <div
        className="divider"
        id="gallery-before"
      />
      <div
        style={Object {}}
      >
        <div
          className="infinite-scroll-component"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "height": "auto",
              "overflow": "auto",
              "position": "relative",
            }
          }
        >
          <div
            className="infinite-scroll"
          >
            <div
              className="preloader-wrapper small active"
            >
              <div
                className="spinner-layer spinner-green-only"
              >
                <div
                  className="circle-clipper left"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="gap-patch"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="circle-clipper right"
                >
                  <div
                    className="circle"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="fixed-action-btn"
  >
    <a
      className="btn-floating btn-large pulse"
      onClick={[Function]}
    >
      <i
        className="material-icons"
      >
        add
      </i>
    </a>
  </div>
</div>
`;
